### DISCUSSION TITLE
Is it possible to do an on_conflict_do_update with unique functional index?

### DISCUSSION THREAD

**Discussion by ChaseLewis** [sqlalchemy/sqlalchemy#8573] [Usage Questions]
I have some tables where I'm planning on forcing a case-insensitive name for.  I can easily do that with a functional index as shown below.

```
class StorageLocation(Base):
    __tablename__ = 'storage_locations'

    storage_location_id = Column(UUID,primary_key=True,server_default=server_default_uuid)
    name = Column(Text,nullable=False)
    created_timestamp = Column(DateTime(timezone=True),nullable=False,server_default=server_default_now)
    type = Column(SmallInteger,nullable=False)
    path = Column(Text,nullable=False)

    __table_args__ = (
        Index("storage_locations_names_idx",func.lower(name),unique=True),
    )
```

However, I'm trying to make the operation a no-op if the location already exists with a on_conflict_do_update since an on_conflict_do_nothing won't return the data appropriately. This would work fine if I did a unique=True on the name column above, but I can't seem to figure out how to specify the index to make this work below. 

```
    statement = (
        insert(models.StorageLocation)
        .values(name=name,type=type,path=path)
        .on_conflict_do_update(index_elements=["storage_locations_names_idx"],set_=dict(created_timestamp=models.StorageLocation.created_timestamp))
        .returning(models.StorageLocation)
    )
```

**Comment by zzzeek** [comment]
that's the correct syntax can you define "not working" ?   what backend is this?   are you able to get the raw SQL statement to do what you want?

**Reply by 4lgorithmic** [reply]
Same issue here.

**Table definition:**

```
    class Profile(Base):

        __table_args__ = (Index('unique_index_profile',
                                'code',
                                'description',
                                'ars_id',
                                unique=True,
                                postgresql_where='ars_id IS NOT NULL'),)

        code: Mapped[str]
        description: Mapped[str]
        ars_id: Mapped[Optional[int]]`
```

From `Base` some other default columns (`id`, `dt_created`, `dt_updated`) are added tot the `Profile`-table. Not important for this case.

**Upsert statement:**

```
 with (Session.begin() as session):
        stmt = insert(model.Profile).values(data)
        stmt = stmt.on_conflict_do_update(index_elements=['unique_index_profile'],
                                          set_=dict(dt_updated=stmt.excluded.dt_updated,
                                                    code=stmt.excluded.code,
                                                    description=stmt.excluded.description,
                                                    ars_id=stmt.excluded.ars_id))

        session.execute(stmt)
```

**Error message:**

`psycopg2.errors.UndefinedColumn: column "unique_index_profile" does not exist`

It seems that only a sequence of column names is supported for the `index_elements`-argument, not custom made `Index` objects added to the table. 

Any ideas?

**Reply by 4lgorithmic** [reply]
Same error message when changing `index_elements=['unique_index_profile']` to `index_elements=model.Profile.__table__.indexes`.

**Reply by zzzeek** [reply]
"unique_index_profile" is an index name.  use the [constraint](https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#sqlalchemy.dialects.postgresql.Insert.on_conflict_do_update.params.constraint) parameter, not `index_where`. https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#specifying-the-target

**Reply by 4lgorithmic** [reply]
Cannot use a constraint, because the argument `postgresql_where=` is not defined for constraints... :-(

You might remember this post: https://github.com/sqlalchemy/sqlalchemy/issues/3161#issuecomment-441925516

I am just looking for a way to upsert some data (Postgres) based on a unique constrain/index over 3 columns that ignores `Null`-values.

**Reply by zzzeek** [reply]
the constraint parameter accepts the index name

>The [Insert.on_conflict_do_update.constraint](https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#sqlalchemy.dialects.postgresql.Insert.on_conflict_do_update.params.constraint) argument is used to specify an index directly rather than inferring it. This **can be the name of** a UNIQUE constraint, a PRIMARY KEY constraint, or **an INDEX**:

**Reply by losthobbit** [reply]
@zzzeek , are you sure?  The documentation doesn't say that, and having tried it I get the error, `constraint "[My index name]" for table "[My table name]" does not exist`

**Reply by zzzeek** [reply]
sqlite and postgresql are different
